protocol/lavasession/consumer_session_manager.go

```diff
@@ -356,32 +394,71 @@ func (csm *ConsumerSessionManager) getValidProviderAddress(ignoredProvidersList
 	csm.lock.RLock()
 	defer csm.lock.RUnlock()
-	currentEpoch = csm.atomicReadCurrentEpoch() // reading the epoch here while locked, to get the epoch of the pairing.
+	currentEpoch := csm.atomicReadCurrentEpoch() // reading the epoch here while locked, to get the epoch of the pairing.
 	if ignoredProviders.currentEpoch < currentEpoch {
 		utils.LavaFormatDebug("ignoredProviders epoch is not the current epoch, resetting ignoredProviders", utils.Attribute{Key: "ignoredProvidersEpoch", Value: ignoredProviders.currentEpoch}, utils.Attribute{Key: "currentEpoch", Value: currentEpoch})
 		ignoredProviders.providers = make(map[string]struct{}) // reset the old providers as epochs changed so we have a new pairing list.
 		ignoredProviders.currentEpoch = currentEpoch
 	}
 
-	providerAddress, err = csm.getValidProviderAddress(ignoredProviders.providers, cuNeededForSession, requestedBlock)
+	// Fetch provider addresses
+	providerAddresses, err := csm.getValidProviderAddresses(ignoredProviders.providers, cuNeededForSession, requestedBlock)
 	if err != nil {
-		utils.LavaFormatError("could not get a provider address", err)
-		return nil, "", 0, err
+		utils.LavaFormatError("could not get a provider addresses", err)
+		return nil, err
 	}
-	consumerSessionsWithProvider = csm.pairing[providerAddress]
-	if err := consumerSessionsWithProvider.validateComputeUnits(cuNeededForSession); err != nil { // checking if we even have enough compute units for this provider.
-		return nil, providerAddress, 0, err // provider address is used to add to temp ignore upon error
+
+	// save how many providers we are aiming to return
+	wantedProviderNumber := len(providerAddresses)
+
+	// Create map to save sessions with providers
+	sessionWithProviderMap = make(SessionWithProviderMap, wantedProviderNumber)
+
+	// Iterate till we fill map or do not have more
+	for {
+		// Iterate over providers
+		for _, providerAddress := range providerAddresses {
+			consumerSessionsWithProvider := csm.pairing[providerAddress]
+			if err := consumerSessionsWithProvider.validateComputeUnits(cuNeededForSession); err != nil {
+				// Add to ignored
+				ignoredProviders.providers[providerAddress] = struct{}{}
+				continue
+			}
+
+			// If no error, add provider session map
+			sessionWithProviderMap[providerAddress] = &SessionWithProvider{
+				SessionsWithProvider: consumerSessionsWithProvider,
+				CurrentEpoch:         currentEpoch,
+			}
+			// Add to ignored
+			ignoredProviders.providers[providerAddress] = struct{}{}
+
+			// If we have enough providers return
+			if len(sessionWithProviderMap) == wantedProviderNumber {
+				return sessionWithProviderMap, nil
+			}
+		}
+
+		// If we do not have enough fetch more
+		providerAddresses, err = csm.getValidProviderAddresses(ignoredProviders.providers, cuNeededForSession, requestedBlock)
+
+		// If error exists but we have providers, return them
+		if err != nil && len(sessionWithProviderMap) != 0 {
+			return sessionWithProviderMap, err
```

Comment:
caller doesnt check this map is there is an error

Reply:
I think that it should check this map even if there is an error. 

This method:
```
// If we do not have enough fetch more
providerAddresses, err = csm.getValidProviderAddresses(ignoredProviders.providers, cuNeededForSession, requestedBlock)
```
can return either, a list of providers (len!=0) or an error. The only error it can return is **_PairingListEmptyError_**. 

The scenario which can happen is that we've requested for example **5** providers, but while validating we removed a few (let's say 2). This is done here https://github.com/lavanet/lava/blob/PRT-514-quorum-select-first-provider/protocol/lavasession/consumer_session_manager.go#L430.

That means that when we exited for loop we do not have the desired number of providers (which was 5), we only have 3 providers. So we would request additional ones https://github.com/lavanet/lava/blob/PRT-514-quorum-select-first-provider/protocol/lavasession/consumer_session_manager.go#L449. 

What can happen is that the second time when we call **getValidProviderAddresses** there are no valid providers. So we have 3 valid from the previous cycle and no more. WHich means that even if we now have an error **_PairingListEmptyError_**, we have some valid providers to return.

Reply:
But something else i see should be fixed here is that if we are returning the sessionWithProviderMap we should not return an error:
Fixed: daaf96276a04ec70c5af95970cea84456fc7ccd1

Reply:
This also impacts this comment https://github.com/lavanet/lava/pull/435#discussion_r1208501673